compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -2468,17 +2468,38 @@ private void createSimpleVarDefStmt(BLangSimpleVariable simpleVariable, BLangBlo
 
     @Override
     public void visit(BLangAssignment assignNode) {
-        boolean fieldAccessLVExpr = assignNode.varRef.getKind() == NodeKind.FIELD_BASED_ACCESS_EXPR;
+        boolean isOptionalFieldAssignment = isOptionalBasicTypeFieldAssignment(assignNode);
         assignNode.varRef = rewriteExpr(assignNode.varRef);
         assignNode.expr = rewriteExpr(assignNode.expr);
         BType castingType = assignNode.varRef.getBType();
-        if (fieldAccessLVExpr) {
+        if (isOptionalFieldAssignment) {
             castingType = types.addNilForNillableAccessType(castingType);
         }
         assignNode.expr = types.addConversionExprIfRequired(rewriteExpr(assignNode.expr), castingType);
         result = assignNode;
     }
 
+    private boolean isOptionalBasicTypeFieldAssignment(BLangAssignment assignNode) {
+        if (assignNode.varRef.getKind() != NodeKind.FIELD_BASED_ACCESS_EXPR) {
+            return false;
+        }
+        BLangFieldBasedAccess fieldAccessNode = (BLangFieldBasedAccess) assignNode.varRef;
+        BType targetType = Types.getImpliedType(fieldAccessNode.expr.getBType());
+        if (targetType.getKind() != TypeKind.RECORD) {
+            return false;
+        }
+        BRecordType recordType = (BRecordType) targetType;
+        BField field = recordType.fields.get(fieldAccessNode.field.value);
+        if (field == null || (field.symbol.flags & Flags.OPTIONAL) != Flags.OPTIONAL) {
+            return false;
+        }
+        BType fieldType = field.getType();
+        return switch (fieldType.getKind()) {
+            case BOOLEAN, FLOAT, STRING, DECIMAL -> true;
+            default -> TypeTags.isIntegerTypeTag(fieldType.tag);
+        };
```

Comment:
```ballerina
public function main() {
    record {|
        string:Char x?;    
    |} r = {x: "A"};
    r.x = ();
}
```

fails with

```
error: casting is not supported from '()' to 'string'
```

You can probably use org.wso2.ballerinalang.compiler.util.TypeTags#isSimpleBasicType with a not nil check.

Reply:
```ballerina
type String string;

public function main() {
    record {|
        String x?;    
    |} r = {x: "A"};
    r.x = ();
}
```

also fails. May need `getImpliedType`.

Reply:
Fixed